rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,379 @@
+allocated up front.
+
+This is probably not ready yet until this PR for `BTreeMap` lands.
+https://github.com/rust-lang/rust/pull/77438
+
+
+### Optimization
+
+With the code as a single dimension array of Bytecode, it could be possible to
+scan the code for patterns and reorganise the Bytecode so it can run in a more
+optimal way.
+
+A lot more thought and research needs to go into this before we can consider
+implementing these changes.
+
+### Optimizing stdlib functions
+
+There is a small number of functions in the stdlib that follow a pattern of
+converting `Bytes` into a `Cow<str>` using `String::from_utf8_lossy`. The
+string is then allocated using `.into_owned()`.
+
+With some shifting around it should be possible to avoid this extra allocation,
+possibly at the cost of a slight reduction in code ergonomics.
+
+The functions that could possibly be optimised are:
+
+`contains`
+`ends_with`
+`starts_with`
+`join`
+`truncate`
+
+## Rationale
+
+Vrl is often identified as a bottleneck in performance tests. Being a key
+component in Vector any optimisation that can be done in Vrl will benefit
+Vector as a whole.
+
+## Drawbacks
+
+Downsides to moving Vrl to use reference counting:
+
+- Using `RefCell` does move the borrow checking to the runtime. Without compile
+  time checks the chances of a panic are much higher.
+
+Downsides to using a Vm:
+
+- The code is much more complex. With an AST that we walk it is fairly apparent
+  what the code will be doing at any point. With a Vm, this is not the case, it
+  is harder to look at the instructions in the Vm and follow back to what part
+  of the Vrl code is being evaluated. We will need to write some extensive
+  debugging tools to allow for decent introspection into the Vm.
+- We lose a lot of safety that we get from the Rust compiler. There will need
+  to be significant fuzz testing to ensure that the code runs correctly under
+  all circumstances.
+- Currently each stdlib function is responsible for evaluating their own
+  parameters. This allows parameters to be lazily evaluated. Most likely with a
+  Vm, the parameters will need to be evaluated up front and the stack passed
+  into the function. This could impact performance.
+
+## Prior Art
+
+- Goscript - An implementation of Go using a bytecode Vm,
+  https://github.com/oxfeeefeee/goscript
+
+- CPython - https://github.com/python/cpython
+
+## Alternatives
+
+This RFC lists a number of improvements that can be made to Vrl, we could do
+all or just some of them. Each of the changes will need to be done with a big
+focus on how much of an improvement is actually possible with each technique.
+Only then will we be able to judge if the extra code complexity is worth the
+improvement.
+
+If we don't do any, Vrl will continue to be a bottleneck. It is also possible
+that Vrl will continue to be a bottleneck after these changes, but hopefully
+just not as significant.
```

Comment:
I don't fully agree with this. I'm still curious how @bruceg's work of processing events in batches is going to help out.

Right now, VRL is single-threaded, and the `remap` transform handles one event at a time. If we are allowed to handle events in batches, we can spin 20 VRL runtimes simultaneously to process 20 events at a time, theoretically getting a 20-fold performance improvement (in practice, it'll be less, but it'll still be a lot).

I've got this nagging feeling in the back of my mind that I'm not entirely convinced we're at a point where we need to invest in any of the bigger items in this RFC, if batch-processing is close to being possible.

I agree, all the items in this RFC will likely contribute to an improved performance of VRL, but some of them also have significant complexity downsides, so it's not entirely clear to me right now if they are worth the investment, or if we should wait for batch processing.

Reply:
For what it's worth pipelining VRL runtimes is something we can already do today and would be worth exploring. Consider that we already have [`ready_chunks`](https://docs.rs/futures/0.3.17/futures/stream/trait.StreamExt.html#method.ready_chunks) in the topology. Each batch -- whether from Bruce's work or pipelining over chunks -- will be limited by the slowest VRL computation, so single-threaded performance still does matter. That's maybe different if we go full-bore streaming and process in parallel, our of order but we don't have plans for that. 

Still, it should be cheap-ish to fiddle with pipelining without waiting for Bruce's work to land.

Reply:
> For what it's worth pipelining VRL runtimes is something we can already do today and would be worth exploring. Consider that we already have [`ready_chunks`](https://docs.rs/futures/0.3.17/futures/stream/trait.StreamExt.html#method.ready_chunks) in the topology.

Interesting! Are you suggesting we experiment by adding a new [`Transform::BatchedFunction`](https://github.com/vectordotdev/vector/blob/43dd342c7ef02a18ddc53620426b394f7386c269/lib/vector-core/src/transform/mod.rs#L10-L18) transform type, that works similar to `Function`, except that it doesn't [iterate over the chunk of events](https://github.com/vectordotdev/vector/blob/43dd342c7ef02a18ddc53620426b394f7386c269/src/topology/builder.rs#L240-L243), but passes them in all at once?

How close is this semantically to the work @bruceg is doing? I'd definitely favour starting up an experiment to benchmark such a change and measure the increase in performance.

> Each batch -- whether from Bruce's work or pipelining over chunks -- will be limited by the slowest VRL computation, so single-threaded performance still does matter.

I fully agree. It's a matter of balancing the required time investment, combined with the increase in implementation complexity, vs. the expected gains. What I'm suggesting is that it _might_ be worth holding off on these changes until VRL is a bit more stable (maybe after 1.0, assuming none of these compiler changes will introduce user-facing breaking changes), if we can get "enough" of a performance boost by processing events in batches, to satisfy all existing use-cases of customers.

I'm suggesting this assuming that the batches processing will give us an x-fold improvement increase, whereas any of the changes suggested here will be in the 10-30% single-threaded performance increase. This assumption might be inaccurate, though.

Reply:
This is related but distinct from the work Bruce is doing. His work is focused on amortizing the internal communication cost of moving events through the topology. What we're talking about here is concurrency within one logical component.

> Are you suggesting we experiment by adding a new [`Transform::BatchedFunction`](https://github.com/vectordotdev/vector/blob/43dd342c7ef02a18ddc53620426b394f7386c269/lib/vector-core/src/transform/mod.rs#L10-L18) transform type, that works similar to `Function`, except that it doesn't [iterate over the chunk of events](https://github.com/vectordotdev/vector/blob/43dd342c7ef02a18ddc53620426b394f7386c269/src/topology/builder.rs#L240-L243), but passes them in all at once?

This is also related but (imo) not the same. Unless VRL has a way to truly operate on multiple events simultaneously, then I don't see any benefit in passing it a batch of events at a time.

Instead, what I think we're discussing is the possibility that the topology could take batches of events and spawn a new task on which it would run them through a VRL runtime, and then collect the results and forward them normally. This would allow us to spawn multiple work tasks concurrently within a single `remap` transform, which no changes to VRL itself.

I definitely agree this is something we should rig up and try, especially since it would be very cheap to do so. The main question is how well it pays off and if there are cases where the overhead of spawning and collecting lead to degradation. Even if we need to branching to only spawn when needed, I'm optimistic it could provide some serious gains and allow us to lean heavily on tokio's existing work-stealing scheduler instead of implementing some additional concurrency mechanism.

Reply:
Yeah, Luke hits the nail on the head I think. It _might_ be that we can pipeline a bunch of parallel VRL instances and operate over a chunk -- however that's originated into the system -- and it _might_ be that VRL is better off truly operating on multiple events at once itself. There are options here and we need to play with them, be ready with an experimental rig to understand their behavior. So, if making a `BatchedFunction` is a cheap spike, give it a shot. This is, I think, also weight against dropping a bunch of time on a new VM _if_ we see other optimization opportunities available to us.

That said, I'd like to push back strongly on the notion that we should hold off on VRL improvements until other streams of work pan out. Today in our experiments VRL is worth -60Mb/s of throughput for `datadog_agent -> remap -> datadog_log`. That's representative and is a deep hole to work out of. It's strongly correlated to VRL's poor single-threaded performance. It's true that we _do_ have other streams of work that are related to improving vector throughput. This RFC represents another stream of work, one that cooperates with the others and gives us some wins even if the others don't pan out in the time we'd like them to. So, we know VRL is a performance problem. Rather than discuss whether we should tackle that problem now in this PR let's discuss how to go about tackling it. We should be interested in finding an ordering of work that land cheap wins up front, delay harder work until we know more about what we have found in the cheap work or have a clearer idea about where our customer pain points are.

Reply:
> This is also related but (imo) not the same. Unless VRL has a way to truly operate on multiple events simultaneously, then I don't see any benefit in passing it a batch of events at a time.

The (my) idea here would be to allow the _remap_ transform to spawn tasks to run multiple VRL runtimes concurrently, not having to make any changes to VRL itself. However, I can see how that isn't much different from having the topology spawn those tasks (it's better, in that we don't have to care about those tasks in the transform implementation itself).

> I'm optimistic it could provide some serious gains and allow us to lean heavily on tokio's existing work-stealing scheduler instead of implementing some additional concurrency mechanism.

This is my thinking as well.

> That said, I'd like to push back strongly on the notion that we should hold off on VRL improvements until other streams of work pan out.

My suggestion would be not to "hold off", but to "switch focus". That is, Stephen has time reserved this quarter to work on "improving the performance of the remap transform". My thesis (unsubstantiated, and potentially incorrect) is that it might be worth investing that time in concurrently processing events in the remap transform, as that will give (again, as an assumption) significantly more benefits in terms of throughput performance, than the 10-30% single-threaded performance improvements we can make with the current proposed work.

I'm also not saying we shouldn't do these improvements at all, and some of them are likely to be useful to do right now, I agree. I'm hesitant to add too much complexity at this stage though, especially if we can get equal or more performance by work that doesn't introduce such complexity, or at least introduces it while benefiting _all_ of Vector topology (e.g. parallel execution).

> It's strongly correlated to VRL's poor single-threaded performance.

I'm still somewhat confused at how we ended up in this situation. When we first shipped VRL, it was equal to, although about ~10% slower than, the native Rust-transforms we replaced (e.g. `parse_json`, `add_field`, etc).

Either those benchmarks were _way_ off the mark, we regressed significantly since then, or VRL programs have become more complicated than were possible at the time, introducing more work a single transform needs to do, as opposed to having multiple transforms each running their own task.

Regardless, I realize this discussion has somewhat drifted off into hypotheticals and has pulled in more people with thoughts on this topic. If there's no logical conclusion to this discussion in sight, it might be worth scheduling a video session to discuss our options and carve a path forward.

I realize @StephenWakely hasn't weighted in here yet, and he's driving this RFC forward, so I'll leave it up to him if he sees value in such a discussion, or if there's enough information here to iterate on the RFC and come to a set of tasks that can get us a good chunk of performance gains over the next couple of months.

Reply:
> > That said, I'd like to push back strongly on the notion that we should hold off on VRL improvements until other streams of work pan out.
> 
> My suggestion would be not to "hold off", but to "switch focus". That is, Stephen has time reserved this quarter to work on "improving the performance of the remap transform". My thesis (unsubstantiated, and potentially incorrect) is that it might be worth investing that time in concurrently processing events in the remap transform, as that will give (again, as an assumption) significantly more benefits in terms of throughput performance, than the single-threaded performance improvements we can make with the current proposed work.

Oh yeah, it's for sure possible. I'd be game to see that approach fiddled with if it's something we can spike and then measure. Even if we declare it out-of-scope for this work it's loosely related to @lukesteensen's work on the topology or something I could take on as an experiment.  This approach will not have a linear or near-linear speed up, for reasons I outlined earlier and considering that vector will be run on low-count CPU systems. Pipelining is absolutely one of the techniques we'll need to bring to bear at some point, though. 
 
> I'm also not saying we shouldn't do these improvements at all, and some of them are likely to be useful to do right now, I agree. I'm hesitant to add too much complexity at this stage though, especially if we can get equal or more performance by work that doesn't introduce such complexity, or at least introduces it while benefiting _all_ of Vector topology (e.g. parallel execution).

Of course. My council for something important like this is to focus on landing things in quick order that provide some immediate benefit, rather that try to land complex stuff with _theoretically_ huge upsides straight out. It seems like [here](https://github.com/vectordotdev/vector/pull/9812#discussion_r740944926) is an example of something that wouldn't be very labor intensive to do and would have an upside, while [this](https://github.com/vectordotdev/vector/pull/9812#discussion_r740939622) might have a benefit and might not but could be complicated to build. A steady trickle of improvements beats a big bang improvement. An approach that would be complicated to build I'd want some convincing proof it'll be worth the time investment, relative to other things. 

> > It's strongly correlated to VRL's poor single-threaded performance.
> 
> I'm still somewhat confused at how we ended up in this situation. When we first shipped VRL, it was equal to, although about ~10% slower than, the native Rust-transforms we replaced (e.g. `parse_json`, `add_field`, etc).
> 
> Either those benchmarks were _way_ off the mark, we regressed significantly since then, or VRL programs have become more complicated than were possible at the time, introducing more work a single transform needs to do, as opposed to having multiple transforms each running their own task.

Ah yeah, hard to say since we haven't been measuring the whole time. Vector _surrounding_ the transforms has gotten significantly faster since VRL was initially introduced, so that's got some part to play in the present situation. One outcome of my work this quarter is an ongoing measurement of vector's throughput, so we'll avoid slipping back unknowingly in the future. Some comfort, at least.

Reply:
All of what you said sounds reasonable to me 👍.

I'm thrilled to see this RFC move forward, but I've yet to be convinced that taking on the VM-related work is a worthwhile investment at this point (not because I don't believe it'll give us gains, but because of the required investment, and the added complexity to the project). I'm not sure what can convince me otherwise, aside from all the other proposed tasks not resulting in any significant gains, or turning out to be too complex in scope (both unlikely).

I'm also eager to see if there's any early spike work we can push to have the remap transform act concurrently on batches of events, it might be out of scope for _this_ RFC, but I'd rather see that work start sooner than later, if nothing else, to prove out potential gains, and weigh its effectiveness relating to the work described in this RFC.

Given the concurrency work is being picked up by others, we could either collaborate on the spike, or slap something together separate from the actual production-ready work, to get those early numbers.

Reply:
> All of what you said sounds reasonable to me +1.
> 
> I'm thrilled to see this RFC move forward, but I've yet to be convinced that taking on the VM-related work is a worthwhile investment at this point (not because I don't believe it'll give us gains, but because of the required investment, and the added complexity to the project). I'm not sure what can convince me otherwise, aside from all the other proposed tasks not resulting in any significant gains, or turning out to be too complex in scope (both unlikely).

FWIW I'm not arguing on this point one way or the other since it's well outside my domain of experience. If folks in the know feel like a VM is not a big lift I'm game, if folks feel like it is I'm glad to see us tackle it later. 

> I'm also eager to see if there's any early spike work we can push to have the remap transform act concurrently on batches of events, it might be out of scope for _this_ RFC, but I'd rather see that work start sooner than later, if nothing else, to prove out potential gains, and weigh its effectiveness relating to the work described in this RFC.
> 
> Given the concurrency work is being picked up by others, we could either collaborate on the spike, or slap something together separate from the actual production-ready work, to get those early numbers.

I like it. This falls into my wheelhouse some, so if we make an issue to do the spike I'll take it on. I'm curious anyway.

Reply:
I just wanted to give some long term context. I'll let you all weigh the benefits/costs, but we eventually want to get to a place where we can push 100mb/s/core with a simple parsing VRL script. Therefore, we will inevitably need to solve single core performance.

Reply:
Good context.

Reply:
So my idea for the focus of this RFC is very much about improving the performance of the VRL runtime - ie the single core performance. Parallelising the transform is I'm sure going to provide other huge wins, but I imagine all the work there is occuring at the topology level and would apply to any transform, not specifically the Remap one.

I'm not suggesting that we have to do all these things straight away, but the RFC is just listing out things we can do to improve the performance. We may choose to do reference counting now, then topology changes and see where that leaves us before moving onto the riskier VM.

But if our goal for single core performance is to get us from 60mb/s -> 100mb/s I think we are going to have to implement the VM, reference counting alone won't cut it. Initial benchmarks with the basic VM indicate that that will get us there.